content/blog/bip158-deep-dive.md

```diff
@@ -0,0 +1,356 @@
+---
+
+title: "Compact Block Filters Deep Dive (BIP 158)"
+subtitle: "A technical explanation of the workings of compact block filters and Golomb-Rice Coding"
+date: 2021-11-13T09:32:00+02:00
+authors:
+  - name: Elle Mouton
+    github: ellemouton
+    twitter: ellemouton
+tags:
+  - "BIP158"
+  - "Compact block filters"
+  - "Bitcoin light client"
+
+images:
+  - "/post-data/compact-block-filters/cover.png"
+
+---
+
+In this post, I will briefly describe the needs of a bitcoin light client and
+why compact block filters satisfy these needs better than Bloom filters do. Then
+I will dive into exactly how compact block filters work and will follow this
+with a step by step guide for constucting such a filter from a testnet block.
+
+#### The purpose of block filters
+
+A bitcoin light client is software that can back a bitcoin wallet without storing the blockchain. This means that it needs to be able to broadcast transactions to the network but most importantly it must be able to pick up when there is a new transaction that is relevant to the wallet it is backing. There are two ways a transaction becomes relevant to a wallet: either it is sending money to the wallet (creating a new output to a wallet address) or it is spending one of the UTXOs that the wallet owns.
+
+#### What was wrong with Bloom filters?
+
+Before [BIP 158] came along, the most widely used method for light clients was to use Bloom filters[^bloomfilters] as described in [BIP 37]. With a bloom filter, you take all the objects you are interested in (script pub keys spent or created), hash them a couple times and add the result of each to a bit map called a Bloom filter. This filter represents what you are interested in. You would then send this filter to a trusted bitcoin node and ask them to send you anything that matches your filter. The problem with this is that it is not very private since you are revealing some information to the bitcoin node you are sending this filter too. They can start getting an idea of the transactions you are interested in as well as the ones you are definitely not interested in. They can also just decide not to send you a transaction that matches the filter. So as you can see, it isn’t great for the light client. But it is also not great for the bitcoin node serving the light client. Each time you send them a filter, they have to load the relevant block from disk and determine which transactions match your filter. You could just spam them with fake filters and effectively DOS them. It takes very little energy to create a filter and lots to respond to it.
```

Comment:
```suggestion
Before [BIP 158] came along, the most widely used method for light clients was to use Bloom filters[^bloomfilters] as described in [BIP 37]. With a bloom filter, you take all the objects you are interested in (script pub keys spent or created), hash them a couple times and add the result of each to a bit map called a Bloom filter. This filter represents what you are interested in. You would then send this filter to a trusted bitcoin node and ask them to send you anything that matches your filter. The problem with this is that it is not very private since you are revealing some information to the bitcoin node you are sending this filter to. They can start getting an idea of the transactions you are interested in as well as the ones you are definitely not interested in. They can also just decide not to send you a transaction that matches the filter. So as you can see, it isn’t great for the light client. But it is also not great for the bitcoin node serving the light client. Each time you send them a filter, they have to load the relevant block from disk and determine which transactions match your filter. You could just spam them with fake filters and effectively DOS them. It takes very little energy to create a filter and lots to respond to it.
```

Ugh, the diff here is 

``` diff
- node you are sending this filter too. 
+ node you are sending this filter to. 
```